http4s/client/src/main/scala/endpoints/http4s/client/Endpoints.scala

```diff
@@ -0,0 +1,268 @@
+import endpoints.Tupler
+import endpoints.Semigroupal
+import endpoints.PartialInvariantFunctor
+import endpoints.Invalid
+import org.http4s.{Request => Http4sRequest, Response => Http4sResponse}
+import org.http4s.Header
+import org.http4s.client.Client
+import org.http4s.Headers
+import endpoints.Validated
+import endpoints.Valid
+import org.http4s.util.CaseInsensitiveString
+import cats.data.Kleisli
+import org.http4s.Status
+import org.http4s.Uri
+
+class Endpoints[F[_]: Sync](val host: Uri, val client: Client[F])
+    extends algebra.Endpoints
+    with EndpointsWithCustomErrors
+    with BuiltInErrors {
+  type Effect[A] = F[A]
+  override def effect: Sync[Effect] = Sync[F]
+}
+
+trait EndpointsWithCustomErrors
+    extends algebra.EndpointsWithCustomErrors
+    with Urls
+    with Methods
+    with StatusCodes {
+
+  type Effect[A]
+  implicit def effect: Sync[Effect]
+
+  def host: Uri
+  def client: Client[Effect]
+
+  type RequestHeaders[A] = (A, Http4sRequest[Effect]) => Http4sRequest[Effect]
+
+  override def emptyRequestHeaders: RequestHeaders[Unit] = (_, req) => req
+
+  override def requestHeader(
+      name: String,
+      docs: Option[String]
+  ): RequestHeaders[String] =
+    (value, req) => req.putHeaders(Header(name, value))
+
+  override def optRequestHeader(
+      name: String,
+      docs: Option[String]
+  ): RequestHeaders[Option[String]] =
+    (value, req) =>
+      value match {
+        case Some(v) => req.putHeaders(Header(name, v))
+        case None    => req
+      }
+
+  implicit def reqHeadersSemigroupal: Semigroupal[RequestHeaders] =
+    new Semigroupal[RequestHeaders] {
+
+      override def product[A, B](fa: RequestHeaders[A], fb: RequestHeaders[B])(
+          implicit tupler: Tupler[A, B]
+      ): RequestHeaders[tupler.Out] =
+        (out, req) => {
+          val (a, b) = tupler.unapply(out)
+          fb(b, fa(a, req))
+        }
+
+    }
+
+  implicit def reqHeadersInvFunctor: InvariantFunctor[RequestHeaders] =
+    new InvariantFunctor[RequestHeaders] {
+
+      override def xmap[A, B](
+          fa: (A, Http4sRequest[Effect]) => Http4sRequest[Effect],
+          f: A => B,
+          g: B => A
+      ): (B, Http4sRequest[Effect]) => Http4sRequest[Effect] =
+        (to, req) => fa(g(to), req)
+
+    }
+
+  type RequestEntity[A] = (A, Http4sRequest[Effect]) => Http4sRequest[Effect]
+
+  implicit def reqEntityInvFunctor: InvariantFunctor[RequestEntity] =
+    new InvariantFunctor[RequestEntity] {
+
+      override def xmap[A, B](
+          fa: (A, Http4sRequest[Effect]) => Http4sRequest[Effect],
+          f: A => B,
+          g: B => A
+      ): (B, Http4sRequest[Effect]) => Http4sRequest[Effect] =
+        (to, req) => fa(g(to), req)
+
+    }
+
+  override def emptyRequest: RequestEntity[Unit] = (_, req) => req
+
+  override def textRequest: RequestEntity[String] =
+    (value, req) => req.withEntity(value)
+
+  type Request[A] = A => Effect[Http4sResponse[Effect]]
```

Comment:
The fact that a `Request[A]` is interpreted as “submitting a request” seems weird. Do you think something like `A => Http4sRequest[Effect]` would work? (and you would do the `submit` in the implementation of `endpoint` only.)

Reply:
I can definitely do this and it makes sense to me, though given that everything is lazy, I don't expect this to make a difference and therefore I can't imagine this being the source of the non-determinism

Reply:
Fixed this in the most recent commit :)